### PR TITLE
rust: Fix linking with C libraries in subdir

### PR DESCRIPTION
**Pull request by xclaesse**
`-l static=` expects a filename and not a relative path, rustc already has `-L` flags to find the library. Weirdly passing a path to `-l dylib=` works fine, but better never do it.

Fixes: #12484

### CODE REVIEW

**Comment by gfxstrand** [comment]
This fixes the Mesa build with Meson 1.3.0rc2.

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1979,6 +1979,15 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
         else:
             verbatim = ''
 
+        def _link_library(libname: str, static: bool, modifiers: T.Optional[T.List[str]] = None, bundle: bool = False):
+            _type = 'static' if static else 'dylib'
```
Comment (bruchar1) [inline mesonbuild/backend/ninjabackend.py:1983]: nitpick: usually, we use _trailing_ underscore in Python to distinguish identifiers from reserved keywords.

https://peps.python.org/pep-0008/#descriptive-naming-styles

Comment (xclaesse) [reply mesonbuild/backend/ninjabackend.py]: done

**Comment by jpakkane** [comment]
It this now ready for merging?

**Comment by xclaesse** [comment]
Yes

**Comment by jpakkane** [comment]
Fails when building under Debian Sid pbuilder:

```
All failures:
  -> rust: 13 external c dependencies    (static=False method=cmake)
  -> rust: 13 external c dependencies    (static=False method=pkg-config)
  -> rust: 13 external c dependencies    (static=False method=system)
  -> rust: 13 external c dependencies    (static=True method=system)
  -> rust: 20 rust and cpp    (b_vscrt=from_buildtype)
  -> rust: 20 rust and cpp    (b_vscrt=md)
  -> rust: 20 rust and cpp    (b_vscrt=mdd)
  -> rust: 20 rust and cpp    (b_vscrt=mt)
  -> rust: 20 rust and cpp    (b_vscrt=mtd)
  -> rust: 20 rust and cpp    (b_vscrt=none)
  -> rust: 20 rust and cpp    (b_vscrt=static_from_buildtype)
```

One of the error messages:

```
rustc -C linker=cc --color=always -C debug-assertions=yes -C overflow-checks=no --crate-type bin -g --crate-name main --emit dep-info=main.d --em
it link=main --out-dir main.p -C metadata=main@exe -Clink-arg=libcpp.a -L. -Clink-arg=-L/usr/lib/gcc/i686-linux-gnu/13 -Clink-arg=-L/usr/lib/gcc/
i686-linux-gnu/13/../../../i386-linux-gnu -Clink-arg=-L/usr/lib/i386-linux-gnu -Clink-arg=-L/usr/lib/gcc/i686-linux-gnu/13/../../../../lib -Clink
-arg=-L/usr/lib -Clink-arg=-L/lib/i386-linux-gnu -Clink-arg=-L/lib/../lib -Clink-arg=-L/usr/lib/../lib -Clink-arg=-L/usr/lib/gcc/i686-linux-gnu/1
3/../../.. -Clink-arg=-L/lib -Clink-arg=-lstdc++ '../test cases/rust/20 rust and cpp/main.rs'
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/usr/lib/rustlib/i686-unknown-linux-gnu/bin:/usr/sbin:/usr/bin:/sbin:/bin" VSLANG="1033" "cc" "-m32" "/tmp/rustcy68F2x
/symbols.o" "main.p/main.main.ffd924bf-cgu.0.rcgu.o" "main.p/main.main.ffd924bf-cgu.1.rcgu.o" "main.p/main.main.ffd924bf-cgu.10.rcgu.o" "main.p/m
ain.main.ffd924bf-cgu.2.rcgu.o" "main.p/main.main.ffd924bf-cgu.3.rcgu.o" "main.p/main.main.ffd924bf-cgu.4.rcgu.o" "main.p/main.main.ffd924bf-cgu.
5.rcgu.o" "main.p/main.main.ffd924bf-cgu.6.rcgu.o" "main.p/main.main.ffd924bf-cgu.7.rcgu.o" "main.p/main.main.ffd924bf-cgu.8.rcgu.o" "main.p/main
.main.ffd924bf-cgu.9.rcgu.o" "main.p/main.4ek1sc2eddz1nvoz.rcgu.o" "-Wl,--as-needed" "-L" "." "-L" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib"
"-Wl,-Bstatic" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libstd-9eb1d1dded28896e.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libpanic_u
nwind-b9975aa5c12eaccf.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libobject-7caf3a8d5b4f043b.rlib" "/usr/lib/rustlib/i686-unknown-linux-g
nu/lib/libmemchr-3b2966b4a4f5d67e.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libaddr2line-34da0594e576609d.rlib" "/usr/lib/rustlib/i686-u
nknown-linux-gnu/lib/libgimli-4e36f80f23111b27.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/librustc_demangle-f2ce595e59ab042e.rlib" "/usr/
lib/rustlib/i686-unknown-linux-gnu/lib/libstd_detect-41160129bd7723c4.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libhashbrown-7f7b27cd909
266e8.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libminiz_oxide-ca2cb570f5eebbb2.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libad
ler-61b1de8adc09be95.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/librustc_std_workspace_alloc-b244d37ff37bedd7.rlib" "/usr/lib/rustlib/i68
6-unknown-linux-gnu/lib/libunwind-b162b23e58e42aac.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libcfg_if-6d47de0badd76ebc.rlib" "/usr/lib/
rustlib/i686-unknown
-linux-gnu/lib/liblibc-abc97f6f160793e3.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/liballoc-909d2cf54addb281.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/librustc_std_workspace_core-2f4f74014cd40ee1.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libcore-c5dc2261107d87e4.rlib" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib/libcompiler_builtins-7ec1d5128420f267.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/usr/lib/rustlib/i686-unknown-linux-gnu/lib" "-o" "main" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "libcpp.a" "-L/usr/lib/gcc/i686-linux-gnu/13" "-L/usr/lib/gcc/i686-linux-gnu/13/../../../i386-linux-gnu" "-L/usr/lib/i386-linux-gnu" "-L/usr/lib/gcc/i686-linux-gnu/13/../../../../lib" "-L/usr/lib" "-L/lib/i386-linux-gnu" "-L/lib/../lib" "-L/usr/lib/../lib" "-L/usr/lib/gcc/i686-linux-gnu/13/../../.." "-L/lib" "-lstdc++"
  = note: /usr/bin/ld: libcpp.a.p/lib.cpp.o: in function `lib_length':
          ./b dccb8dea1f/../test cases/rust/20 rust and cpp/lib.cpp:18:(.text+0xfe): undefined reference to `__stack_chk_fail_local'
          /usr/bin/ld: main: hidden symbol `__stack_chk_fail_local' isn't defined
          /usr/bin/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status

  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: aborting due to previous error
```

**Comment by eli-schwartz** [comment]
That looks like the rust linker is not correctly handling system linking the way the system linker does, and neglects to load e.g. `/usr/lib64/libc_nonshared.a`

**Comment by xclaesse** [comment]
@jpakkane It's weird, none of the flag `-Clink-arg=` seems to get passed down to linker command. Which version of rustc is that, and is there a way to reproducde?

**Comment by eli-schwartz** [comment]
Presumably https://packages.debian.org/source/sid/rustc

**Comment by xclaesse** [comment]
I tried running `./run_project_tests.py --only rust` in a debian sid docker and it passed.

**Comment by jpakkane** [comment]
I tried that also now, but in this case test 13 fails. Rerunning the full packaging test.

**Comment by jpakkane** [comment]
BTW this is on 32 bit x86, not x64.

**Pull request merged**